bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/api/MycroftConnection.java

```diff
@@ -0,0 +1,268 @@
+            handleWrongSession(session, message);
+            return;
+        }
+        logger.trace("{} received raw data: {}", socketName, message);
+
+        try {
+
+            // listeners on message type :
+            BaseMessage mycroftMessage = gson.fromJson(message, BaseMessage.class);
+            Objects.requireNonNull(mycroftMessage);
+            mycroftMessage.message = message;
+
+            // special listener : to any messages
+            Set<MycroftMessageListener<? extends BaseMessage>> listenersAnyToNotify = listeners.get(MessageType.any);
+            if (listenersAnyToNotify != null && !listenersAnyToNotify.isEmpty()) {
+                for (MycroftMessageListener<? extends BaseMessage> listener : listenersAnyToNotify) {
+                    listener.baseMessageReceived(mycroftMessage);
+                }
+            }
+
+            // listener for specific message type
+            Set<MycroftMessageListener<? extends BaseMessage>> listenersToNotify = listeners.get(mycroftMessage.type);
+            if (mycroftMessage.type != MessageType.any && listenersToNotify != null && !listenersToNotify.isEmpty()) {
+                BaseMessage fullMycroftMessage = gson.fromJson(message, mycroftMessage.type.getMessageTypeClass());
+                mycroftMessage.message = message;
+                Objects.requireNonNull(fullMycroftMessage);
+                listenersToNotify.forEach(listener -> listener.baseMessageReceived(fullMycroftMessage));
+            }
+
+        } catch (RuntimeException e) {
+            // we need to catch all processing exceptions, otherwise they could affect the connection
+            logger.warn("{} encountered an error while processing the message {}: {}", socketName, message,
+                    e.getMessage());
+        }
+    }
+
+    @OnWebSocketError
+    public void onError(@Nullable Session session, Throwable cause) {
+
+        if (session == null || !session.equals(this.session)) {
+            handleWrongSession(session, "Connection error: " + cause.getMessage());
+            return;
+        }
+        logger.warn("{} connection error, closing: {}", socketName, cause.getMessage());
+
+        Session storedSession = this.session;
+        if (storedSession != null && storedSession.isOpen()) {
+            storedSession.close(-1, "Processing error");
+        }
+    }
+
+    @OnWebSocketClose
+    public void onClose(Session session, int statusCode, String reason) {
+        if (!session.equals(this.session)) {
+            handleWrongSession(session, "Connection closed: " + statusCode + " / " + reason);
+            return;
+        }
+        logger.trace("{} closed connection: {} / {}", socketName, statusCode, reason);
+        connectionState = ConnectionState.DISCONNECTED;
+        this.session = null;
+        connectionListener.connectionLost(reason);
+    }
+
+    private void handleWrongSession(@Nullable Session session, String message) {
```

Comment:
Same question as above: Is the user able to somehow solve the wrong session problem by himself?
If its not a big issue => reduce to `debug` level.

Reply:
Done